File: 2017/04_HighEntropyPassphrases/test/test_passphrases.py
# ======================================================================
# High-Entropy Passphrases
#   Advent of Code 2017 Day 04 -- Eric Wastl -- https://adventofcode.com
#
# Computer simulation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                t e s t _ p a s s p h r a s e s . p y
# ======================================================================
"Test solver for Advent of Code 2017 day 3, High-Entropy Passphrases"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import unittest

import aoc_04
import passphrases

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------

P1_EXAMPLES_TEXT = """
aa bb cc dd ee
aa bb cc dd aa
aa bb cc dd aaa

! The first and third are valid but the second is not.
"""

# ======================================================================
#                                                        TestPassPhrases
# ======================================================================


class TestPassPhrases(unittest.TestCase):  # pylint: disable=R0904
    """Test PassPhrases object"""

    def test_empty_init(self):
        """Test default PassPhrases creation"""

        # 1. Create default PassPhrases object
        mypass = passphrases.PassPhrases()

        # 2. Make sure it has the default values
        self.assertEqual(mypass.part2, False)
        self.assertEqual(mypass.phrases, [])

        # 3. Check methods
        self.assertEqual(mypass.knt_valid(), 0)

    def test_value_init(self):
        """Test PassPhrases creation with values"""

        # 1. Create default PassPhrases object
        mypass = passphrases.PassPhrases(text=["aa bb cc dd ee",
                                               "aa bb cc dd aa",
                                               "aa bb cc dd aaa"])

        # 2. Make sure it has the default values
        self.assertEqual(mypass.part2, False)
        self.assertEqual(len(mypass.phrases), 3)

        # 3. Check methods
        self.assertEqual(mypass.knt_valid(), 2)

    def test_text_init(self):
        """Test PassPhrases creation from text"""

        # 1. Create default PassPhrases object
        mypass = passphrases.PassPhrases(text=aoc_04.from_text(P1_EXAMPLES_TEXT))

        # 2. Make sure it has the default values
        self.assertEqual(mypass.part2, False)
        self.assertEqual(len(mypass.phrases), 3)

        # 3. Check methods
        self.assertEqual(mypass.knt_valid(), 2)

    def test_is_valid(self):
        """Test if individual phrase is valid"""

        # 1. Create default PassPhrases object
        mypass = passphrases.PassPhrases()

        # 2. Make sure it has the default values
        self.assertEqual(mypass.part2, False)
        self.assertEqual(mypass.phrases, [])

        # 3. Check methods
        self.assertEqual(mypass.is_valid("aa bb cc dd ee"), True)
        self.assertEqual(mypass.is_valid("aa bb cc dd aa"), False)
        self.assertEqual(mypass.is_valid("aa bb cc dd aaa"), True)

    def test_is_valid_two(self):
        """Test if individual phrase is valid"""

        # 1. Create default PassPhrases object
        mypass = passphrases.PassPhrases(part2=True)

        # 2. Make sure it has the default values
        self.assertEqual(mypass.part2, True)
        self.assertEqual(mypass.phrases, [])

        # 3. Check methods
        self.assertEqual(mypass.is_valid("abcde fghij"), True)
        self.assertEqual(mypass.is_valid("abcde xyz ecdab"), False)
        self.assertEqual(mypass.is_valid("a ab abc abd abf abj"), True)
        self.assertEqual(mypass.is_valid("iiii oiii ooii oooi oooo"), True)
        self.assertEqual(mypass.is_valid("oiii ioii iioi iiio"), False)

# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end               t e s t _ p a s s p h r a s e s . p y            end
# ======================================================================

File: 2017/04_HighEntropyPassphrases/test_passphrases.lua
-- ======================================================================
-- HighEntropyPassphrases
--   Advent of Code 2017 Day 04 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                    t e s t _ p a s s p h r a s e s . l u a
-- ======================================================================
-- Test solver for Advent of Code 2017 day 04, HighEntropyPassphrases

-- ----------------------------------------------------------------------
--                                                                require
-- ----------------------------------------------------------------------
local luaunit = require('luaunit')

local passphrases = require('passphrases')

-- ----------------------------------------------------------------------
--                                                              from_text
-- ----------------------------------------------------------------------

function from_text(text)
  -- Break the text into trimed, non-comment lines

  -- 1. We start with no lines
  local result = {}

  -- 2. Loop for lines in the text
  for line in text:gmatch('[^\r\n]+') do

    -- 3. But ignore blank and comment lines
    line = line:gsub("%s*$", "")
    if #line > 0 and "!" ~= line:sub(1, 1) then

      -- 4. Add the line
      table.insert(result, line)
    end
  end

  -- 5. Return a table of cleaned lines
  return result
end

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------
local EXAMPLE_TEXT = [[
aa bb cc dd ee
aa bb cc dd aa
aa bb cc dd aaa
abc bb cc dd cab
]]
local PART_ONE_TEXT = EXAMPLE_TEXT
local PART_TWO_TEXT = EXAMPLE_TEXT

local PART_ONE_RESULT = 3
local PART_TWO_RESULT = 2

-- ======================================================================
--                                                        TestPassphrases
-- ======================================================================

function test_empty_init()
  -- Test the default Passphrases creation

  -- 1. Create default Passphrases object
  local myobj = passphrases:Passphrases()

  -- 2. Make sure it has the default values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 0)
  luaunit.assertEquals(#myobj.words, 0)

end

function test_text_init()
  -- Test the Passphrases object creation from text

  -- 1. Create Passphrases object from text
  local myobj = passphrases:Passphrases({text=from_text(EXAMPLE_TEXT)})

  -- 2. Make sure it has the expected values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(#myobj.text, 4)
  luaunit.assertEquals(#myobj.words, 0)

end

function test_part_one()
  -- Test part one example of Passphrases object

  -- 1. Create Passphrases object from text
  local myobj = passphrases:Passphrases({text=from_text(PART_ONE_TEXT)})

  -- 2. Check the part one result
  luaunit.assertEquals(myobj:part_one({verbose=false, limit=0}), PART_ONE_RESULT)

end

function test_part_two()
  -- Test part two example of Passphrases object

  -- 1. Create Passphrases object from text
  local myobj = passphrases:Passphrases({part2=true, text=from_text(PART_TWO_TEXT)})

  -- 2. Check the part two result
  luaunit.assertEquals(myobj:part_two({verbose=false, limit=0}), PART_TWO_RESULT)

end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
os.exit( luaunit.LuaUnit.run() )

-- ======================================================================
-- end             t e s t _ p a s s p h r a s e s . l u a            end
-- ======================================================================
